ui/vr_manager.js

```diff
@@ -0,0 +1,496 @@
+      switch (spatialInfo.projection) {
+        case 'hequ':
+          unsupported = spatialInfo.hfov != 360;
+          this.vrAsset_ = true;
+          break;
+        case 'fish':
+          this.vrAsset_ = true;
+          unsupported = true;
+          break;
+        default:
+          this.vrAsset_ = false;
+          break;
+      }
+      if (unsupported) {
+        shaka.log.warning('Unsupported VR projection or hfov', spatialInfo);
+      }
+      this.checkVrStatus_();
+    });
+
+    this.loadEventManager_.listen(player, 'nospatialvideoinfo', () => {
+      this.vrAsset_ = false;
+      this.checkVrStatus_();
+    });
+
+    this.loadEventManager_.listen(player, 'unloading', () => {
+      this.vrAsset_ = false;
+      this.checkVrStatus_();
+    });
+
+    this.checkVrStatus_();
+  }
+
+  /**
+   * @override
+   */
+  release() {
+    if (this.loadEventManager_) {
+      this.loadEventManager_.release();
+      this.loadEventManager_ = null;
+    }
+    if (this.eventManager_) {
+      this.eventManager_.release();
+      this.eventManager_ = null;
+    }
+    if (this.vrWebgl_) {
+      this.vrWebgl_.release();
+      this.vrWebgl_ = null;
+    }
+  }
+
+  /**
+   * @param {!shaka.extern.UIConfiguration} config
+   */
+  configure(config) {
+    this.config_ = config;
+    this.checkVrStatus_();
+  }
+
+  /**
+   * Returns if a VR is supported.
+   *
+   * @return {boolean}
+   */
+  isPlayingVR() {
+    return !!this.vrWebgl_;
+  }
+
+  /**
+   * Get the angle of the north.
+   *
+   * @return {?number}
+   */
+  getNorth() {
+    if (!this.vrWebgl_) {
+      shaka.log.alwaysWarn('Not playing VR content');
+      return null;
+    }
+    return this.vrWebgl_.getNorth();
+  }
+
+  /**
+   * Returns the field view.
+   *
+   * @return {?number}
+   */
+  getFieldView() {
+    if (!this.vrWebgl_) {
+      shaka.log.alwaysWarn('Not playing VR content');
+      return null;
+    }
+    return this.vrWebgl_.getFieldView();
+  }
+
+  /**
+   * Set the field view.
+   *
+   * @param {number} fieldView
+   */
+  setFieldView(fieldView) {
+    if (!this.vrWebgl_) {
+      shaka.log.alwaysWarn('Not playing VR content');
+      return;
+    }
+    if (fieldView < 0) {
+      shaka.log.alwaysWarn('Field view should be greater than 0');
+      fieldView = 0;
+    } else if (fieldView > 100) {
+      shaka.log.alwaysWarn('Field view should be less than 100');
+      fieldView = 100;
+    }
+    this.vrWebgl_.setFieldView(fieldView);
+  }
+
+  /**
+   * Toggle stereoscopic mode.
+   */
+  toggleStereoscopicMode() {
+    if (!this.vrWebgl_) {
+      shaka.log.alwaysWarn('Not playing VR content');
+      return;
+    }
+    this.vrWebgl_.togglestereoscopicMode();
+  }
+
+  /**
+   * Increment the yaw in X angle in degrees.
+   *
+   * @param {number} angle
+   */
+  incrementYaw(angle) {
+    if (!this.vrWebgl_) {
+      shaka.log.alwaysWarn('Not playing VR content');
+      return;
+    }
+    if (angle < -359) {
```

Comment:
I'm not sure what the purpose of capping the angle variable here (and in the other increment methods).

This "angle" value is not the final angle post-rotation, it is the amount the angle will be changed by, correct? That's what the method name and description suggests.

I think if I was calling this method and asking for a (say) 540 degree rotation, I would expect it to be effectively rotated by 180 degrees, not 359.

Reply:
I am going to remove the limitation